From 2d31ae0925df36b3a17031bb43d267fb675b4733 Mon Sep 17 00:00:00 2001
From: jiangxin <jiangxin@hygon.cn>
Date: Wed, 25 Aug 2021 14:29:40 +0800
Subject: [PATCH 8/9] anolis: target/i386: csv: load initial image to private
 memory

The initial image of CSV guest should be loaded into private memory
before guest boot.

Add APIs to implement the image load.

Signed-off-by: Xin Jiang <jiangxin@hygon.cn>
Change-Id: I8bd63bd788098c97260c255db1a689751b40dc45
---
 hw/i386/pc_sysfw.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/hw/i386/pc_sysfw.c b/hw/i386/pc_sysfw.c
index c8d9e71b88..d7d5e94da7 100644
--- a/hw/i386/pc_sysfw.c
+++ b/hw/i386/pc_sysfw.c
@@ -37,6 +37,7 @@
 #include "hw/block/flash.h"
 #include "sysemu/kvm.h"
 #include "sev.h"
+#include "csv.h"
 
 #define FLASH_SECTOR_SIZE 4096
 
@@ -263,7 +264,17 @@ void x86_firmware_configure(void *ptr, int size)
             error_report("failed to locate and/or save reset vector");
             exit(1);
         }
+        if (csv_enabled()) {
+            ram_addr_t offset = 0;
+            MemoryRegion *mr;
 
-        sev_encrypt_flash(ptr, size, &error_fatal);
+            mr = memory_region_from_host(ptr, &offset);
+            if (!mr) {
+                error_report("failed to get memory region of flash");
+                exit(1);
+            }
+            csv_load_data(mr->addr + offset, ptr, size, &error_fatal);
+        } else
+            sev_encrypt_flash(ptr, size, &error_fatal);
     }
 }
-- 
2.17.1

